File: Softeer/나무 조경.py
from sys import stdin
n = int(stdin.readline())
board = [[] for _ in range(n)]

for i in range(n):
    board[i] = list(map(int, stdin.readline().split()))
visited = [[0] * n for _ in range(n)]
dx = [-1, 1, 0, 0]
dy = [0, 0 ,-1, 1]
res = 0

def dfs(x, y, depth, sum):
    global res, visited
    visited[x][y] = 1

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if 0<=nx<n and 0<=ny<n and not visited[nx][ny]:
            visited[nx][ny] = 1
            tmp = sum + board[x][y] + board[nx][ny]
            if depth == 3 or (n < 3 and depth == 1):
                res = max(tmp, res)
                visited[nx][ny] = 0
                return

            for i in range(n):
                for j in range(n):
                    if not visited[i][j]:
                        dfs(i, j, depth + 1, tmp)
                        visited[i][j] = 0
            visited[nx][ny] = 0

for i in range(n):
    for j in range(n):
        dfs(i, j, 0, 0)
        visited[i][j] = 0

print(res)

File: Softeer/나무 조경.js
var input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
var n = Number(input[0])
var graph = new Array();
for(let i = 1; i<n + 1;i++){
    graph.push(input[i].split(' ').map(Number));
}

var visited = Array.from(Array(n), () => Array(n).fill(0));
const dx = [-1, 1, 0, 0]
const dy = [0, 0, -1, 1]

var res = 0

const dfs = (depth, x, y, sum) => {
    visited[x][y] = 1

    for(let i = 0;i<4;i++){
        var nx = x + dx[i]
        var ny = y + dy[i]
        if(0<=nx && nx < n && 0<=ny && ny < n && visited[nx][ny] === 0){
            visited[nx][ny] = 1
            var tmp = sum + graph[x][y] + graph[nx][ny]

            if(depth === 3 || (n < 3 && depth === 1)){
                res = Math.max(res, tmp)
                visited[nx][ny] = 0
                return
            }

            for(let i = 0; i < n;i++){
                for(let j = 0; j < n;j++){
                    if(visited[i][j] === 0){
                        dfs(depth + 1, i, j, tmp);
                        visited[i][j] = 0;
                    }
                }
            }
            visited[nx][ny] = 0;
        }
    }
}

for(let x = 0; x < n; x++){
    for(let y = 0; y < n;y++){
        dfs(0, x, y, 0);
        visited[x][y] = 0;
    }
}
console.log(res)